text/0000-router-extensibility.md

```diff
@@ -0,0 +1,165 @@
+
+The DSL implementation should assert that passed options are serializable in order to be forward-compatible with possible future performance optimizations to the route generation and recognition implementation.
+
+`generate(): (match: any) => void;`
+
+This is called by the router and is expected to return a function compatible with the router microlib's `map` method. See the [tildeio/router.js README](https://github.com/tildeio/router.js/) for details.
+
+`mount(_name: string, options: any): void;`
+
+This method is used to include a routable engine in an application. Supported `options` are `as`, `path`, and `resetNamespace`.
+
+There is one additional method on the DSL implementation whose appearance suggests it might be part of the public API: `push`. In fact, it is only invoked by the DSL itself and should be made private by being renamed to `_push`.
+
+Internally, the DSL implementation constructs another instance of the DSL class (i.e. `new DSL()`) when processing callbacks provided to `this.route`). This would be problematic if a developer subclasses the DSL class, so the internals should be changed so that each place we call `new DSL` [here](https://github.com/emberjs/ember.js/blob/master/packages/ember-routing/lib/system/dsl.js), we instead do:
+
+```js
+let DSL = this.constructor;
+new DSL()
+```
+
+## How we teach this
+
+It is not expected that beginner-level Ember developers would need to use these hooks. There is nothing made possible by this feature that can't be accomplished without it, albeit more verbosely and spread out across route definitions.
+
+As a result, this feature would merit only passing mention in the guides and rely on API docs as the primary teaching mechanism to addon authors or advanced developers.
+
+Addons which leverage these hooks will bear the "teaching burden" of how a developer needs to modify their `router.js` and routes file to work with the addon.
+
+## Drawbacks
+
+Allowing additional options passed to `this.route` to have meaning implies that any future options used by Ember core run the risk of name collision with application or addon code.
+
+Allowing alternate DSL implementations makes future changes to the DSL API carry more risk.
+
+In general, more public API surface area means more complexity, so the win of providing flexibility to advanced users via a lower-level API might be outweighed by the API surface area increase.
+
+Historically, `Ember.Route` instances were supposed to be "stateless". This makes them somewhat more stateful. Is that a problem?
+
+## Alternatives
+
+This risk of option name collisions could be eliminated by requiring that userland options be nested under hash named `routeOptions` or similar. However, this would negatively affects the readability of the the code and limits the expressiveness of the DSL, which is significant part of the goal of this RFC.
```

Comment:
Maybe it's worth flushing out more how this would limit the expressiveness. I think one big drawback of this approach is not name collisions but a regular ember user installing an addon and wondering why certain properties are now appearing in their routes. The connection between the router map and the route is definitely something that can be learned, but could be baffling. 

On option would be to keep the options hash in the router dsl flat, like it is already, but to just put those options on the route under a `routeOptions` key?